Add Domain, SSL, ELB part to IaSQL on EC2 blog post (#2265)
iasql/iasql: site/blog/unlisted/iasql-on-ec2.mdx

This will make the setup very similar to IaSQL production setup.

<img src='/img/iasql-on-ec2/production-setup.png' />

#### Import Domain In Route53

Let's start with adding our domain to Route53. To be able to manage Route53 resources in IaSQL, we'll first need to install the corresponding module:
```sql
SELECT iasql_install('aws_route53');
```
Alright, now let's `INSERT` a row into the `hosted_zone` table to direct IaSQL engine to create a new Route53 hosted zone for us.

:::note
To know which tables are supported by the `aws_route53` module, you can visit [`aws_route53` module documentation](https://iasql.com/docs/modules/aws/aws_route53/).
:::

```sql
SELECT iasql_begin();

INSERT INTO hosted_zone (domain_name)
VALUES ('<your-domain>');

SELECT iasql_commit();
```

:::note
Simple `INSERT`, right? But wait! What are those `iasql_begin` and `iasql_commit` functions called before and after our insert command?

Quick answer is that if you need your changes to be applied now instead of eventually in the future, you'll wrap the changes between an `iasql_begin` and an `iasql_commit`. For more info on what they do, visit [this page](https://iasql.com/docs/transaction/).
:::

> Now that we have created our hosted zone, we need to set our domain's nameservers to the ones that Route53 provides us. For each public hosted zone that you create, Amazon Route53 [automatically creates](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/SOA-NSrecords.html) a name server (NS) record and a start of authority (SOA) record. So we should log into the domain registrar, and set the name server values to the four name server values that Route53 has set. To get those values, we can simply run a `SELECT` query:

Reviewer: Nit

```suggestion
Now that we have created our hosted zone, we need to set our domain's nameservers to the ones that Route53 provides us. For each public hosted zone that you create, Amazon Route53 [automatically creates](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/SOA-NSrecords.html) a name server (NS) record and a start of authority (SOA) record. So we need to log into our domain registrar and set the name server values to the four name server values that Route53 has set. How this is done exactly varies from registrar to registrar, but we can get those NS values with a `SELECT` query:
```

```sql
SELECT record as name_server
FROM resource_record_set
WHERE parent_hosted_zone_id = (SELECT id FROM hosted_zone WHERE domain_name = '<your-domain>')
  AND record_type = 'NS';
```

> After updating the name servers in the domain settings, we need to wait for those values to be propagated. This can happen in minutes, or even up to 48 hours depending on the registrar.

Reviewer: Providing some more actionable advice here to know when you can move on in the tutorial.

````suggestion
After updating the name servers in the domain settings, we need to wait for those values to be propagated. This can happen in minutes, or even up to 48 hours depending on the registrar.

You can find out which NS records are accessible from your DNS with the [`dig` command](https://linux.die.net/man/1/dig):

```sh
dig <your_domain_name> ns
```

It should look something like this when ready:

```sh
$ dig skybase.dev ns

; <<>> DiG 9.18.1-1ubuntu1.3-Ubuntu <<>> skybase.dev ns
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27469
;; flags: qr rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;skybase.dev.		                IN	NS

;; ANSWER SECTION:
skybase.dev.	        21600	IN	NS	ns-1243.awsdns-27.org.
skybase.dev.	        21600	IN	NS	ns-1944.awsdns-51.co.uk.
skybase.dev.	        21600	IN	NS	ns-49.awsdns-06.com.
skybase.dev.	        21600	IN	NS	ns-697.awsdns-23.net.

;; Query time: 60 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Wed Mar 29 20:48:04 CDT 2023
;; MSG SIZE  rcvd: 179
```
````

#### Issue An SSL Certificate For Our Domain
Having the domain setup in Route53, it's easy to use IaSQL's `certificate_request` RPC to issue a new certificate for it. If you try to go with this procedure through AWS console, you'll need to manually add the CNAME records from the AWS ACM to your hosted zone. But this RPC is implemented in a way that automagically [sets the DNS records](https://github.com/iasql/iasql/blob/v0.1.7/src/modules/aws_acm/rpcs/request.ts#L177-L188), [waits for the certificate to be issued](https://github.com/iasql/iasql/blob/v0.1.7/src/modules/aws_acm/rpcs/request.ts#L197), and then [removes those validation DNS records](https://github.com/iasql/iasql/blob/v0.1.7/src/modules/aws_acm/rpcs/request.ts#L200).

```sql
-- SELECT iasql_install('aws_acm'); -- no need for this because aws_route53 depends on it and it's automatically installed

> SELECT certificate_request('<your-domain>', 'DNS', default_aws_region(), '{}');

Reviewer: Note for a future feature addition: there should be a simplification of this function to not need the last 2 (or even 3?) arguments. These seem like reasonable defaults to me.

Reviewer 2: agreed

```sql
INSERT INTO load_balancer (load_balancer_name, scheme, load_balancer_type, ip_address_type, region)
VALUES ('iasql-lb', 'internet-facing', 'application', 'ipv4', default_aws_region());

INSERT INTO target_group (target_group_name, target_type, protocol, port)
VALUES ('iasql', 'instance', 'HTTPS', 9876);

INSERT INTO listener (load_balancer_id, port, protocol, target_group_id)
VALUES ((SELECT id FROM load_balancer WHERE load_balancer_name = 'iasql-lb'), 443, 'HTTPS',
        (SELECT id FROM target_group WHERE target_group_name = 'iasql'));
```

And do not forget to assign the security group to our load balancer:
```sql
INSERT INTO load_balancer_security_groups (load_balancer_id, security_group_id)
SELECT (SELECT id FROM load_balancer WHERE load_balancer_name = 'iasql-lb'),
       (SELECT id FROM security_group WHERE group_name = 'iasql-lb');
```

The only step remaining is to allow incoming traffic to our instance on port 9876 only from our load balancer:
```sql
UPDATE security_group_rule
SET source_security_group = (SELECT id FROM security_group WHERE group_name = 'iasql-lb'),
    cidr_ipv4             = NULL
WHERE cidr_ipv4 = '<my-ip>';
```

> (we're basically updating the previous rule allowing direct access from our IP to allow access from load balancer security group)

Reviewer: I think this makes it a bit clearer *exactly* what is going on.

```suggestion
(we're basically updating the previous rule that allowed direct access from our IP to no longer allow direct access and instead only allow access from load balancer security group)
```

Waiting for enough time for the above changes to be propagated to the cloud, the load balancer should be up and listening on a DNS name:
```sql
SELECT dns_name FROM load_balancer WHERE load_balancer_name = 'iasql-lb'
```

> Making a CNAME record pointing to that load balancer will make IaSQL working under our domain name:

Reviewer: ```suggestion
Finally, we will create a CNAME record pointing to that load balancer that will make IaSQL working under our domain name:
```

  AND record_type = 'NS';
```
After updating the name servers in the domain settings, we need to wait for those values to be propagated. This can happen in minutes, or even up to 48 hours depending on the registrar.

:::note
You can find out which NS records are accessible from your DNS with the [`dig` command](https://linux.die.net/man/1/dig):

```sh
dig <your-domain> ns
```

It should look something like this when ready:

```sh
$ dig skybase.dev ns

; <<>> DiG 9.18.1-1ubuntu1.3-Ubuntu <<>> skybase.dev ns
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27469
;; flags: qr rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;skybase.dev.		                IN	NS

;; ANSWER SECTION:
skybase.dev.	        21600	IN	NS	ns-1243.awsdns-27.org.
skybase.dev.	        21600	IN	NS	ns-1944.awsdns-51.co.uk.
skybase.dev.	        21600	IN	NS	ns-49.awsdns-06.com.
skybase.dev.	        21600	IN	NS	ns-697.awsdns-23.net.

;; Query time: 60 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Wed Mar 29 20:48:04 CDT 2023
;; MSG SIZE  rcvd: 179
```
:::

> #### Issue An SSL Certificate For Our Domain

Reviewer 2: what does this look like? the fact that this grayed out makes me think if might not properly parsing the sections

Reviewer: Docusaurus' MDX markdown is *not* Github-style markdown, so I do highly recommend actually pulling the branch, and running:

```
cd site
yarn
yarn start:local
```

and navigating to the blogpost to see what it'll actually look like. That's what I've been doing on this PRs.

Author: <img width="786" alt="image" src="https://user-images.githubusercontent.com/5357435/229236517-eebd203c-a637-462d-98b4-035159601002.png">
